[Merged by Bors] - feat(linear_algebra): the determinant of an endomorphism

Vierkantor opened the pull request:
`linear_map.det` is the determinant of an endomorphism, which is defined independent of a basis. If there is no finite basis, the determinant is defined to be equal to `1`.

This approach is inspired by `linear_map.trace`.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

- [x] depends on: #7631
- [x] depends on: #7633

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

Vierkantor committed:
feat(linear_algebra): the determinant of an endomorphism

`linear_map.det` is the determinant of an endomorphism, which is defined independent of a basis. If there is no finite basis, the determinant is defined to be equal to `1`.

This approach is inspired by `linear_map.trace`.

eric-wieser commented on src/linear_algebra/determinant.lean:
```diff
@@ -52,6 +55,149 @@ variables {M' : Type*} [add_comm_group M'] [module R M']
+then { to_fun := @linear_map.det_aux _ _ _ _ H.some_spec.some_spec.some _ _ _ H.some_spec.some,
+       map_one' := @det_aux_one _ _ _ _ H.some_spec.some_spec.some _ _ _ H.some_spec.some,
+       map_mul' := @det_aux_mul _ _ _ _ H.some_spec.some_spec.some _ _ _ H.some_spec.some }
+else 1
```
Using `finset` saves you a lot of underscores:

```suggestion
protected def det : (M →ₗ[A] M) →* A :=
if H : ∃ (s : finset M), nonempty (basis (s : set M) A M)
then { to_fun := linear_map.det_aux H.some_spec.some,
       map_one' := det_aux_one H.some_spec.some,
       map_mul' := det_aux_mul H.some_spec.some }
else 1
```
Does it make things more painful later?

Vierkantor replied:
With `set`, you can directly supply `basis.reindex_range (b : basis ι R M) : basis (set.range b) R M`, as in `det_to_matrix` (which is the most important way to unfold `linear_map.det` in practice). With `finset` there would be some awkwardness involving giving an equiv between `set.finite.to_finset (set.fintype_range b)` and `set.range b` itself.

A neat solution overall, though not quite so easy, is to define `basis.reindex_finset_range` that gives a `finset`-indexed basis.

Vierkantor replied:
(I should get back to fixing #7575, then this would be even nicer!)

eric-wieser commented on src/linear_algebra/determinant.lean:
```diff
@@ -52,6 +55,149 @@ variables {M' : Type*} [add_comm_group M'] [module R M']
+See `linear_map.det` for a version that does not depend on a choice of basis.
+-/
+protected def det_aux (b : basis ι A M) (f : M →ₗ[A] M) : A :=
+matrix.det (linear_map.to_matrix b b f)
```
Is it worth giving this the monoid_hom structure immediately?
```suggestion
/-- The determinant of an endomorphism given a basis.
See `linear_map.det` for a version that does not depend on a choice of basis.
-/
def det_aux (b : basis ι A M) : (M →ₗ[A] M) →* A :=
(monoid_hom.of (matrix.det)).comp (to_matrix_alg_equiv b : (M →ₗ[A] M) →* matrix ι ι A)
```
It breaks dot notation in a few places below, but otherwise seems to work fine. You might be able to eliminate a handful of proofs too.

eric-wieser replied:
You could even fold `det_aux_eq` into the same definition:
```lean
/-- The determinant of an endomorphism given a basis.

See `linear_map.det` for a version that populates the basis non-computably
-/
def det_aux : trunc (basis ι A M) → (M →ₗ[A] M) →* A :=
trunc.lift
  (λ b : basis ι A M,
    (monoid_hom.of (matrix.det)).comp (to_matrix_alg_equiv b : (M →ₗ[A] M) →* matrix ι ι A))
  (λ b c, monoid_hom.ext $ det_to_matrix_eq_det_to_matrix b c)
```

Vierkantor replied:
I didn't go for fancy types since I expect `det_aux` won't be used a lot anyway. In any case, the `monoid_hom` did help a bit with conciseness. It felt like `trunc` didn't have enough API to make the change worth it on its own, but here it is anyway :)

Vierkantor committed:
Better use of `decidable_eq` in `linear_map.det`

The definition of `linear_map.det` still doesn't require `decidable_eq`
but any lemmas unfolding it to `matrix.det` do.

Discussed here: https://github.com/leanprover-community/mathlib/pull/7667#discussion_r635647706

Vierkantor committed:
Use fancier type for `det_aux`

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

Vierkantor committed:
Fix

eric-wieser commented on src/linear_algebra/determinant.lean:
```diff
@@ -95,6 +95,113 @@ end
+-- Auxiliary lemma, the `simp` normal form goes in the other direction
+-- (using `linear_map.det_to_matrix`)
+lemma det_eq_det_to_matrix_of_finite_set [decidable_eq M]
+  {s : set M} (b : basis s A M) (hs : fintype s) (f : M →ₗ[A] M) :
```
Why `fintype s` and not `s.finite`?

Vierkantor replied:
Because (the `matrix` type,) `matrix.det` and `linear_map.to_matrix` need the index to be a fintype.

eric-wieser replied:
Does making it a `[]` argument make anything worse?

Vierkantor replied:
In the end no, since it's only used in `det_to_matrix` where the specific instance is inferrable.

Vierkantor committed:
Replace explicit param with instance param

eric-wieser commented on src/linear_algebra/determinant.lean:
```diff
@@ -95,6 +95,112 @@ end
+
+See `linear_map.det` for a version that populates the basis non-computably.
+-/
+def det_aux : trunc (basis ι A M) → (M →ₗ[A] M) →* A :=
```
Does it makes sense to call this `det_of_basis`? Or is that more confusing?

Vierkantor replied:
I see what you're getting at, though I'm not convinced: if you have a basis, you get a lot more affordances when you directly go to `det (linear_map.to_matrix b b f)`.

eric-wieser replied:
I'm now realizing that this `trunc` stuff isn't nearly as useful as I'd hoped - as stated, this only helps for basis' indexed by the same type.Even if you change it to something monstrous like `trunc (Σ (ι : Type w') [fintype ι] [decidable_eq ι], basis ι A M)`, it still doesn't quantify over multiple universes, so you still can't prove `det_aux_def'` via properties of `trunc`.

Maybe it's still worth having anyway either because a) you've already written it! b) it's common to have different bases from `fin 3 -> R`.

Vierkantor replied:
I have expanded the docstring with some notes on the pros and cons of the `trunc` parameter, sounds good?

eric-wieser commented on src/linear_algebra/determinant.lean:
```diff
@@ -95,6 +95,112 @@ end
+  linear_map.det_aux (trunc.mk b) f = matrix.det (linear_map.to_matrix b b f) :=
+rfl
+
+attribute [irreducible] linear_map.det_aux
```
A comment for why this is here might be good.

Vierkantor replied:
IIRC, it's just that `det_aux` is a relatively complicated term and unfolding is not worth it, since non-refl lemmas like `det_aux_def'` are much more powerful.
```suggestion
-- Discourage the elaborator from unfolding `det_aux` and producing a huge term.
attribute [irreducible] linear_map.det_aux
```

eric-wieser commented on src/linear_algebra/determinant.lean:
```diff
@@ -95,6 +95,112 @@ end
+
+end
+
+attribute [irreducible] linear_map.det
```
Same for here.

Vierkantor replied:
```suggestion
-- Discourage the elaborator from unfolding `det` and producing a huge term.
attribute [irreducible] linear_map.det
```

Vierkantor committed:
Explain why `det_aux` and `det` are irreducible

Vierkantor commented on src/linear_algebra/determinant.lean:
```diff
@@ -95,6 +95,114 @@ end
+
+/-- The determinant of an endomorphism given a basis.
+
+See `linear_map.det` for a version that populates the basis non-computably.
```
```suggestion
See `linear_map.det` for a version that populates the basis non-computably.

Although the `trunc (basis ι A M)` parameter makes it slightly more convenient to switch bases,
there is no good way to generalize over universe parameters, so we can't fully state in `det_aux`'s
type that it does not depend on the choice of basis. Instead you can use the `det_aux_def'` lemma,
or avoid mentioning a basis at all using `linear_map.det`.
```

Vierkantor committed:
Explain what `trunc` can and cannot do for `det_aux`
